File: Drawers/src/PlotUtility.cc
#include "../interface/PlotUtility.h"
#include "THStack.h"
#include "TH1D.h"
#include "TCanvas.h"
#include "TLegend.h"
#include "TFile.h"
#include "TLatex.h"
#include "TSystem.h"
#include "TTreeFormula.h"
#include <math.h>
#include <iostream>
#include <fstream>

using namespace std;

TString sanitize(TString sIn) { 
  TString sOut = sIn; 
  return sOut.ReplaceAll(".","_").ReplaceAll("(","").ReplaceAll(")","").ReplaceAll("/","Over").ReplaceAll("*","Times").ReplaceAll("+","Plus");
}

PlotUtility::PlotUtility(double x, double y):
  HistogramDrawer(x,y)
{
  processes = new Process*[nProcesses];
  owned = new bool[nProcesses];
  for (unsigned int iP=0; iP!=nProcesses; ++iP) {
    processes[iP]=0;
    owned[iP]=false;
  }
}

PlotUtility::~PlotUtility() {
  for (unsigned int iP=0; iP!=nProcesses; ++iP) {
    if (owned[iP])
      delete processes[iP];
  }
  delete processes;
  delete owned;
}

void PlotUtility::AddProcess(Process *p) {
  ProcessType iP = p->processtype;
  if (owned[iP])
    delete processes[iP];
  owned[iP] = false;
  processes[iP] = p;
  order.push_back((unsigned int)iP);
}

void PlotUtility::AddFile(ProcessType pt,TString filePath,TString name) {
  if (processes[pt]==NULL) {
    owned[pt] = true;
    Process *tmp = new Process(name,pt);
    processes[pt] = tmp;
  }
  processes[pt]->AddFile(filePath);
}

void PlotUtility::ClearProcess(ProcessType pt) {
  if (owned[pt])
    delete processes[pt];
  processes[pt]=0;
  owned[pt] = false;
  order.clear(); // could probs implement this better
}

void PlotUtility::Reset(bool clearPlotLabels) {
  // overriding because there was something I wanted
  // to implement here, but I have forgotten what
  // that is
  HistogramDrawer::Reset(clearPlotLabels);
}

TString convertName(TString filename) {
  TString tmpName(filename);
  tmpName.ReplaceAll("/","Over");
  tmpName.ReplaceAll("*","Times");
  tmpName.ReplaceAll("+","Plus");
  tmpName.ReplaceAll("-","Minus");
  tmpName.ReplaceAll("TMath::","");
  tmpName.ReplaceAll(")","");
  tmpName.ReplaceAll("(","");
  tmpName.ReplaceAll(".","_");
  return tmpName; 
}

void divideBinWidth(TH1D *h) {
  // I trust this more than TH1::Scale(1,"width")
  int nBins = h->GetNbinsX();
  for (int iB=1; iB!=nBins+1; ++iB) {
    float currentVal = h->GetBinContent(iB);
    float currentErr = h->GetBinError(iB);
    float binWidth = h->GetBinWidth(iB);
    h->SetBinContent(iB,currentVal/binWidth);
    h->SetBinError(iB,currentErr/binWidth);
  }
}

void PlotUtility::DrawAll(TString outDir) {
  if (!legend)
    InitLegend();
  TFile *fOut = new TFile(outDir+"hists.root","UPDATE");
	if (fOut->IsZombie()) {
		fOut->Close();
		fOut = new TFile(outDir+"hists.root","RECREATE");
	}
	//TFile *fBuffer = new TFile(TString::Format("/tmp/buffer_%i.root",gSystem->GetPid()).Data(),"RECREATE");
  TFile *fBuffer = new TFile(TString::Format("/uscmst1b_scratch/lpc1/3DayLifetime/buffer_%i.root",gSystem->GetPid()).Data(),"RECREATE");
  fBuffer->cd();
  if (order.size()==0) {
    for (unsigned int iP=0; iP!=nProcesses; ++iP) 
      order.push_back(iP);
  }
  

  for (unsigned int iP : order) {
    Process *p = processes[iP];
    if (p==NULL)
      continue;

    for (TTree *t : p->trees) {
      t->SetBranchStatus("*",0);

      turnOnBranches(t,eventnumber);
      turnOnBranches(t,cut.GetTitle());
      if (p->processtype!=kData && p->useCommonWeight) {
        turnOnBranches(t,mcWeight.GetTitle());
      }
      for(Distribution *d : distributions) {
        turnOnBranches(t,d->name);
      }
      turnOnBranches(t,p->additionalCut.GetTitle());
      turnOnBranches(t,p->additionalWeight.GetTitle());
      for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
        turnOnBranches(t,mcWeightUp[iS].GetTitle());
        turnOnBranches(t,mcWeightDown[iS].GetTitle());
			}
    }
  }

  map<Distribution*,PlotWrapper> pws;
  // first we declare all the necessary histograms
  for (Distribution *d: distributions) {
    PlotWrapper pw;
    TH1D *hbase=0;
    if (d->binEdges==0)
      hbase = new TH1D("hbase","hbase",d->nBins,d->min,d->max);
    else
      hbase = new TH1D("hbase","hbase",d->nBins,d->binEdges);
    // hbase->Sumw2();
    TString tmpname = convertName(d->filename);
    // these are the systematic shifts
    for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
      TH1D *hsystup = (TH1D*)hbase->Clone(TString::Format("h_%s_%s_Up",
                                                          tmpname.Data(),systNames[iS].Data()));
      TH1D *hsystdown = (TH1D*)hbase->Clone(TString::Format("h_%s_%s_Down",
                                                          tmpname.Data(),systNames[iS].Data()));
      hsystup->Sumw2(); hsystdown->Sumw2();
      hsystup->SetLineColor(systColors[iS]); hsystup->SetLineWidth(2);
      hsystdown->SetLineColor(systColors[iS]); hsystdown->SetLineWidth(2);
      pw.hSystUp.push_back(hsystup);
      pw.hSystDown.push_back(hsystdown);
    }
    // one histogram for each process
    for (unsigned int iP : order) {
      Process *p = processes[iP];
      if (p==NULL)
        continue;
      TH1D *h = (TH1D*)hbase->Clone(TString::Format("h_%s_%s",tmpname.Data(),p->name.Data()));
      h->Sumw2();
      pw.histos[p] = h;
      pw.ownedHistos.push_back(h);
    }
    pws[d] = pw;
    hbase->Delete();
  }

  // now we loop through each process
  TimeReporter tr("PlotUtility::DrawAll",true);
  tr.Start();
  for (unsigned int iP : order) {
    Process *p = processes[iP];
    if (p==NULL)
      continue;
    // build the cut and weight formulae
    TCut finalCut = cut;
    TCut finalWeight = "1";
    if (p->processtype!=kData && p->useCommonWeight)
      finalWeight = mcWeight;
    if (p->processtype<=kSignal3 && p->processtype!=kData)
      finalWeight *= TCut(TString::Format("%f",signalScale));
    if (eventmod!=0) {
      if (p->processtype==kData)
        finalCut += TCut(TString::Format("(%s%%%i)==0",eventnumber.Data(),eventmod).Data());
      else
        finalWeight *= TCut(TString::Format("%f",1./eventmod).Data());
    }
    finalCut += p->additionalCut;
    finalWeight *= p->additionalWeight;

    // loop through each subprocess
    for (TTree *drawTree : p->trees) {
      TTreeFormula fcut(finalCut.GetTitle(),finalCut.GetTitle(),drawTree); fcut.SetQuickLoad(true); fcut.GetNdata();
      TTreeFormula fweight(finalWeight.GetTitle(),finalWeight.GetTitle(),drawTree); fweight.SetQuickLoad(true); fweight.GetNdata();

      // build the systematic shift weights
      vector<TTreeFormula*> fsystups, fsystdowns;
      vector<double> syst_up_weights(systNames.size());
      vector<double> syst_down_weights(systNames.size());
			if (p->processtype!=kData && p->useCommonWeight) {
				for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
					TCut upWeight = mcWeightUp[iS];
					upWeight *= p->additionalWeight;
					if (eventmod!=0)
						upWeight *= TCut(TString::Format("%f",1./eventmod).Data());
					TTreeFormula *fup = new TTreeFormula(upWeight.GetTitle(),upWeight.GetTitle(),drawTree); fup->SetQuickLoad(true);
					fsystups.push_back(fup);

					TCut downWeight = mcWeightDown[iS];
					downWeight *= p->additionalWeight;
					if (eventmod!=0)
						downWeight *= TCut(TString::Format("%f",1./eventmod).Data());
					TTreeFormula *fdown = new TTreeFormula(downWeight.GetTitle(),downWeight.GetTitle(),drawTree); fdown->SetQuickLoad(true);
					fsystdowns.push_back(fdown);
				}
			} else {
				for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
					fsystups.push_back(&fweight);
					fsystdowns.push_back(&fweight);
				}
			}
      
      // build the distribution formulae
      for (auto *d : distributions) {
        delete pws[d].tf;
        pws[d].tf = new TTreeFormula(d->name,d->name,drawTree);
        pws[d].tf->SetQuickLoad(true);
      }
      
      // cut, evaluate, and fill
      unsigned int nEntries = drawTree->GetEntries();
      for (unsigned int iE=0; iE!=nEntries; ++iE) {
        drawTree->GetEntry(iE);
        if (!fcut.EvalInstance())
          continue;
        double weight = fweight.EvalInstance();
				if (p->processtype!=kData) {
					for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
						syst_up_weights[iS] = fsystups[iS]->EvalInstance();
						syst_down_weights[iS] = fsystdowns[iS]->EvalInstance();
					}
				}
        for (auto *d : distributions) {
          PlotWrapper &pw = pws[d];
          TH1D *h = pw.histos[p];
          double val = pw.tf->EvalInstance();
          double minX = d->min, maxX = d->max;
          if (doOverflow && val>maxX)
            val = h->GetBinCenter(d->nBins);
          else if (doUnderflow && val<minX)
            val = h->GetBinCenter(1);
          h->Fill(val,weight);
					if (p->processtype!=kData) {
						for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
							pw.hSystUp.at(iS)->Fill(val,syst_up_weights[iS]);
							pw.hSystDown.at(iS)->Fill(val,syst_down_weights[iS]);
						}
					}
        }
      }

			if (p->processtype!=kData && p->useCommonWeight) {
				for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
					delete fsystups[iS]; delete fsystdowns[iS];
					fsystups[iS]=0; fsystdowns[iS]=0;
				}
			}

    }
    tr.TriggerEvent(p->name.Data());
  }

  // now that all histograms are filled,
  // draw each plot
  for (Distribution *d : distributions) {
    PlotWrapper &pw = pws[d];

    bool doScaleBins = d->binEdges!=NULL;

    if (d->name=="1") {
      double bgTotal=0, bgErr=0;
      double sigTotal=0, sigErr=0;
      double dataTotal=0, dataErr=0;
      std::ofstream fyields; fyields.open(outDir+"yields.txt");

      TString stable = TString::Format("%-25s | %15s \\pm %15s","Process","Yield","Error");
      fyields << stable << std::endl;  PInfo("PlotUtility::Dump",stable);

      stable = "=================================================================";
      fyields << stable << std::endl;  PInfo("PlotUtility::Dump",stable);

      for (unsigned int iP : order) {
        Process *p = processes[iP];
        if (p==NULL)
          continue;
        TH1D *h = pw.histos[p];
        double integral=0,error=0;
        integral = h->IntegralAndError(1,h->GetNbinsX(),error);
        TString syield = TString::Format("%-25s | %15f \\pm %15f",p->name.Data(),integral,error);
        fyields << syield.Data() << std::endl;
        PInfo("PlotUtility::Dump",syield);
        if (p->processtype==kData) {
          dataTotal=integral; dataErr=error;
        } else if (p->processtype<=kSignal3) {
          sigTotal += integral;
          sigErr += pow(error,2); 
        } else {
          bgTotal += integral;
          bgErr += pow(error,2);
        }
      }
      fyields << stable << std::endl;  PInfo("PlotUtility::Dump",stable);

      sigErr = sqrt(sigErr);
      bgErr = sqrt(bgErr);
      
      TString syield = TString::Format("%-25s | %15f \\pm %15f","MC(bkg)",bgTotal,bgErr);
      fyields << syield.Data() << std::endl;      PInfo("PlotUtility::Dump",syield);
      syield = TString::Format("%-25s | %15f \\pm %15f","MC(sig)",sigTotal,sigErr);
      fyields << syield.Data() << std::endl;      PInfo("PlotUtility::Dump",syield);
      syield = TString::Format("%-25s | %15f \\pm %15f","Data",dataTotal,dataErr);
      fyields << syield.Data() << std::endl;      PInfo("PlotUtility::Dump",syield);
      fyields << stable << std::endl;             PInfo("PlotUtility::Dump",stable);

      syield = TString::Format("S/B=%.3f, S/sqrtB=%.3f",sigTotal/bgTotal,sigTotal/sqrt(bgTotal));
      fyields << syield.Data() << std::endl; PInfo("PlotUtility::Dump",syield);

      fyields.close();
    } 
    for (unsigned int iP : order) {
      Process *p = processes[iP];
      if (p==NULL)
        continue;
      TH1D *h = pw.histos[p];
      if (doScaleBins)
        divideBinWidth(h);
      if (d->maxY>-999)
        h->SetMaximum(d->maxY);
      if (d->minY<999)
        h->SetMinimum(d->minY);
      h->GetXaxis()->SetTitle(d->xLabel);
      h->GetYaxis()->SetTitle(d->yLabel);
      h->SetTitle("");
      if (!doStack && p->processtype!=kData)
        h->SetLineWidth(2);
      if (p->processtype<=kSignal3 && p->processtype!=kData && signalScale!=1)
        AddHistogram(h,TString::Format("%.1f#times%s",signalScale,p->name.Data()).Data(),p->processtype);
      else  
        AddHistogram(h,p->name.Data(),p->processtype);  
    }
    for (unsigned int iS=0; iS!=systNames.size(); ++iS) {
      if (doScaleBins) {
        divideBinWidth(pw.hSystUp[iS]);
        divideBinWidth(pw.hSystDown[iS]);
      }
      AddSystematic(pw.hSystUp[iS],"hist",systNames[iS]);
      AddSystematic(pw.hSystDown[iS],"hist");
    }
    TString tmpname = convertName(d->filename);
    Logy(false);
    HistogramDrawer::Draw(outDir,tmpname);
    tmpname += "_logy";
    legend->Clear();
    Logy(true);
    HistogramDrawer::Draw(outDir,tmpname);
    Reset(false);
    for (auto h : pw.ownedHistos) {
      fOut->WriteTObject(h);
      delete h;
    }
    for (auto h : pw.hSystUp) {
      fOut->WriteTObject(h);
      delete h;
    }
    for (auto h : pw.hSystDown) {
      fOut->WriteTObject(h);
      delete h;
    }
    pw.ownedHistos.clear();
    pw.hSystUp.clear();
    pw.hSystDown.clear();
    delete pw.tf; pw.tf=0;
  }

  distributions.clear();
  fOut->Close(); // do not write or cloned trees will get saved
  fBuffer->Close();
  delete fBuffer;
  delete fOut;

  
} 

File: Drawers/python/plot_utility.py
#!/usr/bin/env python

import ROOT as root
import numpy as np
from array import array
from PandaCore.Tools.Misc import *
from PandaCore.Tools.Load import Load
from PandaCore.Tools.root_interface import read_files, draw_hist
from os import getenv, system
from pprint import pprint

Load('HistogramDrawer')

tree_name = 'events'

class Process():
    def __init__(self, name, pt, custom_color=root.nProcesses):
        self.name = name
        self.process_type = pt
        if (custom_color==root.nProcesses):
            self.color = pt
        else:
            self.color = custom_color
        # public config - defaults
        self.dashed = False
        self.dotted = False
        self.use_common_weight = True 
        self.use_common_cut = True
        self.additional_cut = '1==1'
        self.additional_weight = '1'
        self.tree_name = tree_name
        self.ratio = False
        # private stuff
        self.__files = []
        return
    def add_file(self, fpath):
        self.__files.append(fpath)
    def read(self, variables, weights, cut):
        return read_files(filenames = self.__files, 
                          branches = variables+weights, 
                          cut = cut, 
                          treename = self.tree_name)

def convert_name(n):
    rn = str(n)
    rn = rn.replace("/", "Over");    
    rn = rn.replace("*", "Times");   
    rn = rn.replace("+", "Plus");    
    rn = rn.replace("-", "Minus");   
    rn = rn.replace("TMath::", "");  
    rn = rn.replace(")", "");        
    rn = rn.replace("(", "");        
    rn = rn.replace(".", "_");       
    return rn


def divide_bin_width(h):
    nb = h.GetNbinsX()
    for ib in xrange(1, nb+1):
        val = h.GetBinContent(ib)
        err = h.GetBinError(ib)
        width = h.GetBinWidth(ib)
        h.SetBinContent(ib, val/width)
        h.SetBinError(ib, err/width)


def fix_underflow(h):
    h.SetBinContent(1, h.GetBinContent(0)+h.GetBinContent(1))

def fix_overflow(h):
    nbins = h.GetNbinsX()
    h.SetBinContent(nbins, h.GetBinContent(nbins+1)+h.GetBinContent(nbins))


class Distribution():
    def __init__(self, name, bin_list, is_variable, xlabel, ylabel, filename=None, ybounds=None):
        self.name = name
        if name=='1':
            # override
            is_variable = False
            bin_list = (0, 2, 1)
        self.is_variable = is_variable
        self.xlabel = xlabel
        self.ylabel = ylabel
        self.filename = convert_name(filename if filename else name)
        self.ybounds = ybounds
        self.histograms = {} # so we can access all hists of this distribution
        self.systs = {}
        self.calc_chi2 = False
        if is_variable:
            self.bins = array('f', bin_list)
            self.nbins = len(bin_list)-1
            self.hbase = root.TH1D('hbase_%s'%(self.filename), 
                                   '', self.nbins, self.bins)
        else:
            self.xbounds = (bin_list[0], bin_list[1])
            self.nbins = bin_list[2]
            self.hbase = root.TH1D('hbase_%s'%(self.filename), 
                                   '', self.nbins, 
                                   self.xbounds[0], self.xbounds[1])
    def generate_hist(self, label):
        self.histograms[label] = self.hbase.Clone('h_%s_%s'%(self.filename, convert_name(label)))
        self.histograms[label].Sumw2()
        return self.histograms[label]
    def generate_syst(self, label):
        hup = self.hbase.Clone('h_%s_%s_Up'%(self.filename, convert_name(label)))
        hdown = self.hbase.Clone('h_%s_%s_Down'%(self.filename, convert_name(label)))
        hup.Sumw2(); hdown.Sumw2()
        self.systs[label] = (hup, hdown)
        return (hup, hdown)

def FDistribution(name, lo, hi, nbins, xlabel, ylabel, filename=None, ybounds=None):
    return Distribution(name=name, bin_list=(lo, hi, nbins), is_variable=False, 
                        xlabel=xlabel, ylabel=ylabel, filename=filename, ybounds=ybounds)

def VDistribution(name, bins, xlabel, ylabel, filename=None, ybounds=None):
    return Distribution(name=name, bin_list=bins, is_variable=True, 
                        xlabel=xlabel, ylabel=ylabel, filename=filename, ybounds=ybounds)

class Systematic():
    def __init__(self, name, weight_up, weight_down, color=1):
        self.name = name
        self.weight_up = weight_up
        self.weight_down = weight_down
        self.color = color
        self.hists = None
    def generate_weight(self, central, is_up):
        replacement = self.weight_up if is_up else self.weight_down
        if type(replacement) == str:
            return tTIMES(central, replacement)
        else:
            shift = str(central)
            for k, v in replacement.iteritems():
                shift = shift.replace(k, v)
            return shift

class PlotUtility():
    def __init__(self):
        self.canvas = root.HistogramDrawer()
        members = [x for x in dir(self.canvas) if ('__' not in x)]
        for m in members:
            if callable(getattr(self.canvas, m)):
                self.__getPlotMethod(m)
            else:
                self.__getPlotMember(m)
        self.__processes = [] 
        self.__distributions = []
        self.__systematics = []
        # public configuration - defaults
        self.plot_label = '#it{CMS Preliminary}'
        self.do_underflow = True
        self.do_overflow = True
        self.signal_scale = 1
        self.cut = None
        self.mc_weight = None
        self.eventmod =  0
        self.eventnumber = 'eventNumber'
        return
    def __getPlotMethod(self, x):
        method = getattr(self.canvas, x)
        setattr(self, x,  lambda *args : method(*args))
    def __getPlotMember(self, x):
        member = getattr(self.canvas, x)
        setattr(self, x,  member)
    def add_process(self, p):
        self.__processes.append(p)
    def add_distribution(self, d):
        self.__distributions.append(d)
    def add_systematic(self, name, up, down, color=1):
        self.__systematics.append(Systematic(name, up, down, color))
    def Draw(self):
        # override the thing that's bound in __init__
        PError('plot_utility.PlotUtility.Draw', 'Not implemented!')
    def draw_all(self, outdir):
        if not self.canvas.HasLegend():
            self.canvas.InitLegend()

        f_out = root.TFile(outdir+'hists.root', 'UPDATE')
        if f_out.IsZombie():
            f_out.Close()
            f_out = root.TFile(outdir+'hists.root', 'RECREATE')
        f_buffer_path = '/tmp/%s/buffer_%i.root'%(getenv('USER'), root.gSystem.GetPid())
        f_buffer = root.TFile(f_buffer_path, 'RECREATE')
        f_buffer.cd()

        variables = []
        for dist in self.__distributions:
            for syst in self.__systematics:
                syst.hists = dist.generate_syst(syst.name)

            for proc in self.__processes:
                dist.generate_hist(proc.name)

            variables.append(dist.name)

        # loop through each process
        for proc in self.__processes:
            # figure out the nominal weight and cut strings
            final_weight = '1'
            final_cut = self.cut
            if (proc.process_type!=root.kData and proc.use_common_weight):
                final_weight = self.mc_weight
            if (proc.process_type<=root.kSignal3 and proc.process_type!=root.kData):
                final_weight = tTIMES(final_weight, str(self.signal_scale))
            if self.eventmod:
                if (proc.process_type==root.kData):
                    final_cut = tAND(final_cut, '(%s%%%i)==0'%(self.eventnumber, self.eventmod))
                else:
                    final_weight = tTIMES(final_weight, str(1./self.eventmod))
            final_cut = tAND(final_cut, proc.additional_cut)
            final_weight = tTIMES(final_weight, proc.additional_weight)

            weight_map = {'nominal' : final_weight}
            weights = [final_weight]
            if proc.process_type!=root.kData:
                for syst in self.__systematics:
                    if proc.use_common_weight:
                        up_weight = syst.generate_weight(final_weight, True)
                        down_weight = syst.generate_weight(final_weight, False)
                    else:
                        continue
                    weight_map['%s_Up'%(syst.name)] = up_weight
                    weight_map['%s_Down'%(syst.name)] = down_weight
                    weights.append(up_weight)
                    weights.append(down_weight)
            
            xarr = proc.read(variables, weights, final_cut)

            for dist in self.__distributions:
                vals = xarr[dist.name]
                weights_nominal = xarr[weight_map['nominal']]
                draw_hist(hist = dist.histograms[proc.name], 
                          xarr = xarr, 
                          fields = (dist.name, ), 
                          weight = weight_map['nominal'])
                if proc.process_type!=root.kData:
                    for syst in self.__systematics:
                        if proc.use_common_weight:
                            weights_up = weight_map['%s_Up'%syst.name]
                            weights_down = weight_map['%s_Down'%syst.name]
                        else:
                            weights_up = weight_map['nominal']
                            weights_down = weight_map['nominal']
                        draw_hist(hist = dist.systs[syst.name][0], 
                                  xarr = xarr, 
                                  fields = (dist.name, ), 
                                  weight = weights_up)
                        draw_hist(hist = dist.systs[syst.name][1], 
                                  xarr = xarr, 
                                  fields = (dist.name, ), 
                                  weight = weights_down)

        # everything is filled,  now draw the histograms!
        for dist in self.__distributions:
            if dist.name=="1":
                totals = {'bg':0,  'sig':0,  'data':0}
                errs = {'bg':0,  'sig':0,  'data':0}
                table_format = '%-25s | %15f +/- %15f'

                table = ['%-25s | %15s +/- %15s'%('Process', 'Yield', 'Stat. unc.')]
                table.append("=================================================================")

                for proc in self.__processes:
                    h = dist.histograms[proc.name]
                    integral = h.GetBinContent(1)
                    error = h.GetBinError(1)
                    table.append(table_format%(proc.name, integral, error))
                    if proc.process_type==root.kData:
                        proc_label = 'data'
                    elif proc.process_type<=root.kSignal3:
                        proc_label = 'sig'
                    else:
                        proc_label = 'bg'
                    totals[proc_label] += integral
                    errs[proc_label] += pow(error, 2)

                for k, v in errs.iteritems():
                    errs[k] = np.sqrt(v)

                table.append("=================================================================")
                table.append(table_format%('MC(bkg)', totals['bg'], errs['bg']))
                table.append(table_format%('MC(sig)', totals['sig'], errs['sig']))
                table.append(table_format%('Data', totals['data'], errs['data']))

                table.append("=================================================================")
                if totals['bg']:
                    table.append('S/B=%.3f, S/sqrtB=%.3f'%(totals['sig']/totals['bg'], 
                                                           totals['sig']/np.sqrt(totals['bg'])))

                with open(outdir+'yields.txt', 'w') as fyields:
                    fyields.write('\n'.join(table))

                for t in table:
                    PInfo('plot_utility.PlotUtility.Dump', t)

            h_unscaled = {'data':None, 'mc':None} # used for chi2 calc
            for proc in self.__processes:
                h = dist.histograms[proc.name]
                if self.do_overflow:
                    fix_overflow(h)
                if self.do_underflow:
                    fix_underflow(h)
                if proc.process_type==root.kData:
                    h_unscaled['data'] = h.Clone()
                elif proc.process_type>root.kSignal3:
                    if not h_unscaled['mc']:
                        h_unscaled['mc'] = h.Clone()
                    else:
                        h_unscaled['mc'].Add(h)
                if dist.is_variable:
                    divide_bin_width(h)
                if dist.ybounds:
                    h.SetMinimum(ybounds[0])
                    h.SetMaximum(ybounds[1])
                h.GetXaxis().SetTitle(dist.xlabel)
                h.GetYaxis().SetTitle(dist.ylabel)
                h.SetTitle('')
                if self.canvas.IsStack():
                    h.SetLineWidth(2)
                if proc.dashed:
                    h.SetLineStyle(2)
                elif proc.dotted:
                    h.SetLineStyle(3)

                add_hist_args = [h]
                if (proc.process_type<=root.kSignal3 
                    and proc.process_type!=root.kData 
                    and self.signal_scale!=1):
                    add_hist_args.append( '%.1f#times%s'%(self.signal_scale, proc.name) )
                else:
                    add_hist_args.append( proc.name )
                add_hist_args.append( proc.process_type )
                if proc.color != proc.process_type:
                    add_hist_args.append( proc.color )
                else:
                    add_hist_args.append( -1 )
                add_hist_args += ['', proc.ratio]
                self.canvas.AddHistogram(*add_hist_args)

                f_out.WriteTObject(h, h.GetName(), "overwrite")
            for syst in self.__systematics:
                hup, hdown = dist.systs[syst.name]
                for h in [hup, hdown]:
                    if self.do_overflow:
                        fix_overflow(h)
                    if self.do_underflow:
                        fix_underflow(h)
                    if dist.is_variable:
                        divide_bin_width(h)
                    h.SetLineWidth(3)
                    h.SetLineColor(syst.color)
                self.canvas.AddSystematic(hup, 'hist', syst.name)
                self.canvas.AddSystematic(hdown, 'hist')
                f_out.WriteTObject(hup, hup.GetName(), "overwrite")
                f_out.WriteTObject(hdown, hdown.GetName(), "overwrite")

            # output the canvas
            if dist.calc_chi2:
                p = h_unscaled['data'].Chi2Test(h_unscaled['mc'],'UW')
                self.canvas.AddPlotLabel('P(#chi^{2}|NDoF)=%.3g'%(p),0.6,0.5,False,42,.04)
            self.canvas.Logy(False)
            self.canvas.Draw(outdir, dist.filename)
            self.canvas.ClearLegend()
            if dist.calc_chi2:
                p = h_unscaled['data'].Chi2Test(h_unscaled['mc'],'UW')
                self.canvas.AddPlotLabel('P(#chi^{2}|NDoF)=%.3g'%(p),0.6,0.5,False,42,.04)
            self.canvas.Logy(True)
            self.canvas.Draw(outdir, dist.filename+'_logy')

            self.canvas.Reset(False)

        f_out.Close()
        f_buffer.Close()
        system('rm %s'%f_buffer_path)
